lib/Conversion/FIRRTLToRTL/LowerToRTL.cpp

```diff
@@ -1520,258 +1719,99 @@ LogicalResult FIRRTLLowering::visitDecl(MemOp op) {
-    Value en;
-    Value addr;
-    Value mask;
-    Value data;
-    Value rd_en;
-    Value rd_addr;
-    Value rd_mask;
-    Value rd_data;
-  };
-
-  // Create a register for the memory.
-  Value reg =
-      builder.create<sv::RegOp>(rtl::UnpackedArrayType::get(dataType, depth),
-                                builder.getStringAttr(memName.str()));
-
-  // Some helpers.
-  auto buildPAssign = [&](Value dest, Value value) {
-    builder.create<sv::PAssignOp>(dest, value);
-  };
-
-  // Track pipeline registers that were added. These need to be
-  // randomly initialized later.
-  SmallVector<Value> pipeRegs;
+  FirMemory memSummary = analyzeMemOp(op);
 
   // Process each port in turn.
-  SmallVector<std::pair<Identifier, MemOp::PortKind>> ports = op.getPorts();
-
-  for (size_t i = 0, e = ports.size(); i != e; ++i) {
-    auto portName = ports[i].first.str();
-    auto port = op.getResult(i);
-
-    // Do not lower ports if they aren't used.
-    if (port.use_empty())
-      continue;
-
-    auto portBundleType =
-        port.getType().cast<FIRRTLType>().getPassiveType().cast<BundleType>();
-
-    // Create wires for all the memory ports
-    llvm::StringMap<Value> portWires;
-    for (BundleType::BundleElement elt : portBundleType.getElements()) {
-      auto fieldType = lowerType(elt.type);
-      if (fieldType.isInteger(0)) {
-        portWires.insert({elt.name.getValue(), Value()});
-        continue;
-      }
-      auto name =
-          (Twine(memName) + "_" + portName + "_" + elt.name.getValue()).str();
-      auto fieldWire = createTmpWireOp(fieldType, name);
-      portWires.insert({elt.name.getValue(), fieldWire});
-    }
-
-    // Now that we have the wires for each element, rewrite any subfields to
-    // use them instead of the subfields.
-    while (!port.use_empty()) {
-      auto portField = cast<SubfieldOp>(*port.user_begin());
-      portField->dropAllReferences();
-      (void)setLowering(portField, portWires[portField.fieldname()]);
-    }
-
-    // Don't emit anything more for zero bit data values.
-    if (dataType.isInteger(0))
-      continue;
+  SmallVector<Type, 8> resultTypes;
+  SmallVector<Value, 8> readOperands;
+  SmallVector<Value, 8> rwOperands;
+  SmallVector<Value, 8> writeOperands;
+  DenseMap<Operation *, size_t> returnHolder;
+
+  size_t readCount = 0;
+  size_t writeCount = 0;
+  size_t readwriteCount = 0;
+
+  // Memories return multiple structs, one for each port, which means we have
+  // two layers of type to split appart.
+  for (size_t i = 0, e = op.getNumResults(); i != e; ++i) {
+    auto portName = op.getPortName(i).getValue();
+    auto portKind = *op.getPortKind(i);
+
+    auto &portKindNum =
+        portKind == MemOp::PortKind::Read
+            ? readCount
+            : portKind == MemOp::PortKind::Write ? writeCount : readwriteCount;
+    auto addInput = [&](SmallVectorImpl<Value> &operands, StringRef field,
+                        size_t width) {
+      auto portType =
```

Comment:
@darthscsi I guess what I'm saying is to add something like 
```suggestion
width = std::max(1U, width);
```
Up here at the top, so all the uses of `width` have this effect. Is that correct? Seems to work locally for me.

Reply:
Correctly handling the subfieldop requires knowing the original width.  That could be recomputed from the type or it could be passed along.  I don't see a compelling way to do it either way.

Reply:
Hmm ok I guess that makes sense. I guess there is more to this than I was understanding.

If it helps, here is a boiled down testcase for a memory with depth = 1, which should be suitable for dropping into lower-to-rtl.mlir: [test.txt](https://github.com/llvm/circt/files/6307350/test.txt)

That will hit this assertion on the current revision: https://github.com/llvm/circt/blob/260179d2123875b0ad295474fbf39f616586154f/lib/Conversion/FIRRTLToRTL/LowerToRTL.cpp#L1358

Reply:
Thanks, fixed.

Reply:
Awesome, thanks.